### PR TITLE
Force merge rest api support wait_for_completion

### PR DESCRIPTION
**Pull request by mushao999**
- Backgroud:
Force merge action is a very costly action. It may take servial hours to run for big indices. But current  force merge rest api do not support wait_for_completion parameter, we can only close connection or wait for timeout.
related issue: [#80129](https://github.com/elastic/elasticsearch/issues/80129) 
- What this PR do
  - This PR reference form reindex api from code to doc.  
  - code: Add wait_for_completion parameter to force merege rest api
  - test: Add a simple yaml rest test for wait_for_completion parameter of force merge rest api
  - doc: Add wait_for_completion parameter in force merge api doc

### CODE REVIEW

**Review by sethmlarson — DISMISSED** [review]
LGTM from a spec perspective, one nit:

File: docs/reference/indices/forcemerge.asciidoc [hunk 1]
```diff
@@ -52,11 +52,22 @@ performance.
 [[forcemerge-blocks]]
 ===== Blocks during a force merge
 
-Calls to this API block until the merge is complete. If the client connection
+Calls to this API block until the merge is complete (unless request contains
+wait_for_completion=false, which is default ture). If the client connection
```
[the remark this answers is not in the document]
Comment (mushao999) [reply docs/reference/indices/forcemerge.asciidoc]: Thanks , typo fixed.

**Comment by mushao999** [comment]
@sethmlarson Could you please me merge this PR, thank you .

**Comment by sethmlarson** [comment]
I only reviewed the API spec change, not the functionality itself. Needs a different engineer to review first.

**Comment by mushao999** [comment]
> I only reviewed the API spec change, not the functionality itself. Needs a different engineer to review first.

@sethmlarson Thanks, is there anyone we can ping to help review the functionality part?

**Comment by mushao999** [comment]
hi @leehinman , Could you please review this PR?

**Comment by mushao999** [comment]
@andreidan Could you please reivew this PR?

**Comment by andreidan** [comment]
@elasticmachine ok to test

File: server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestForceMergeAction.java [hunk 2]
```diff
@@ -40,7 +47,34 @@ public RestChannelConsumer prepareRequest(final RestRequest request, final NodeC
...
+        }
+        /*
+         * Let's try and validate before forking so the user gets some error. The
+         * task can't totally validate until it starts but this is better than
+         * nothing.
+         */
+        ActionRequestValidationException validationException = mergeRequest.validate();
+        if (validationException != null) {
+            throw validationException;
+        }
```
Comment (andreidan) [inline server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestForceMergeAction.java:63]: `validate` is already called by the transport infrastructure. I don't think we need to create this "branch only" case where we sometimes validate the request based on the `wait_for_completion` parameter. 
I think we should remove this section.

File: server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestForceMergeAction.java [hunk 3]
```diff
@@ -40,7 +47,34 @@ public RestChannelConsumer prepareRequest(final RestRequest request, final NodeC
...
+        return sendTask(
+            client.getLocalNodeId(),
+            client.executeLocally(ForceMergeAction.INSTANCE, mergeRequest, LoggingTaskListener.instance())
+        );
```
Comment (andreidan) [inline server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestForceMergeAction.java:67]: nit: shall we move this block in the `else` block so we don't have "leaking" returns?

**Comment by andreidan** [comment]
@elasticmachine update branch

**Review by andreidan — CHANGES_REQUESTED** [review]
Thanks for the contribution @mushao999 

This is looking good. I've left a couple of minor comments.

File: server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestForceMergeAction.java (hunk 2 above)
Comment (mushao999) [reply server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestForceMergeAction.java:63]: That's right, `TransportAction#execute(...)` will also do the validation. I will remove the validation here, thanks.

File: server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestForceMergeAction.java (hunk 3 above)
Comment (mushao999) [reply server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestForceMergeAction.java:67]: ok, I will move this block in the else block.

**Comment by mushao999** [comment]
@andreidan I've update the code following your suggestion.

**Comment by mushao999** [comment]
@andreidan By the way, my previous code refers to the `AbstractBaseReindexRestHandler`. So that class also has duplicated validation and retrun lacking issue. Shall we fix it in this PR, or open a new PR to do that?

**Comment by andreidan** [comment]
> By the way, my previous code refers to the `AbstractBaseReindexRestHandler`. So that class also has duplicated validation and retrun lacking issue. Shall we fix it in this PR, or open a new PR to do that?

@mushao999 Thanks for iterating on this! I've taken a closer look at that request validation. I think keeping it makes sense (irrespective of the duplication in `TransportAction#execute`) because when we do NOT wait for completion we create the task and change the listener to a `LoggingTaskListener` which will *not* report the possible validation errors that the execution will detect (in `TransportAction#execute`) but will merely log the validation errors (and the users will just see a `200` response).

Can you please add it back, in the `else` branch above this line? https://github.com/elastic/elasticsearch/pull/80463/files#diff-d9fbdc33f49322766f016b5ab5728c5a14737214b1ea72094fa3991572f139b4R53  

Thanks and sorry for not seeing this initially. This'll be ready to merge after this!

**Comment by mushao999** [comment]
@andreidan Request validation is back.  And what about the return lacking in the `AbstractBaseReindexRestHandler`?

**Comment by mushao999** [comment]
@andreidan  Ready to force

**Comment by mushao999** [comment]
@andreidan Ready to merge, thanks

**Review by andreidan — APPROVED** [review]
LGTM, thanks for iterating on this @mushao999

**Comment by andreidan** [comment]
> And what about the return lacking in the AbstractBaseReindexRestHandler

**Pull request merged**